source/server/http/admin.cc

```diff
@@ -544,19 +544,50 @@ Http::Code AdminImpl::handlerRuntime(absl::string_view url, Http::HeaderMap& res
                                      Buffer::Instance& response) {
   Http::Code rc = Http::Code::OK;
   const Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
-  const auto& entries = server_.runtime().snapshot().getAll();
-  const auto pairs = sortedRuntime(entries);
 
+  const auto& layers = server_.runtime().snapshot().getAllLayers();
   if (params.size() == 0) {
-    for (const auto& entry : pairs) {
-      response.add(fmt::format("{}: {}\n", entry.first, entry.second.string_value_));
+    std::set<std::string> keys;
+    static constexpr absl::string_view key_header = "KEY";
+
+    // Determine the maximum width for each column.
+    std::vector<size_t> column_widths;
+    size_t key_column_width = key_header.size();
+    for (const auto& layer : layers) {
+      column_widths.push_back(layer->name().size());
+      for (const auto& kv : layer->values()) {
+        key_column_width = std::max(key_column_width, kv.first.size());
+        column_widths.back() = std::max(column_widths.back(), kv.second.string_value_.size());
+        keys.insert(kv.first);
+      }
+    }
+
+    static const std::string cell_format = "{:<{}}\t";
+    // Add header row.
+    response.add(fmt::format(cell_format, key_header, key_column_width));
+    for (size_t i = 0; i < layers.size(); i++) {
+      response.add(
+          fmt::format(cell_format, StringUtil::toUpper(layers[i]->name()), column_widths[i]));
+    }
+    response.add("\n");
+
+    // Add a row for each key.
+    for (const auto& key : keys) {
+      response.add(fmt::format(cell_format, key, key_column_width));
+      for (size_t i = 0; i < layers.size(); i++) {
+        const auto& layer_values = layers[i]->values();
+        const auto value_it = layer_values.find(key);
+        const auto& value = value_it != layer_values.end() ? value_it->second.string_value_ : "";
+        response.add(fmt::format(cell_format, value, column_widths[i]));
+      }
+      response.add("\n");
     }
   } else {
     if (params.begin()->first == "format" && params.begin()->second == "json") {
+      // TODO: Reviewers, do we want to proto-ize this now? Or continue manually constructing JSON?
```

Comment:
I included the table for human consumption when debugging, which as I understand it is a main motivation for exposing the layers in the first place. 

I don't know much about mtools, but it seems that you'd have to structure the data in a specific way to make it human-readable when pretty-printing, and even then it might be hard to parse visually. That is, rows and columns as implemented here constitute precisely how a human wants to see this data.

That's why I'm opposed to deleting the table now, which raises my question of whether JSON is materially useful until proto/tooling around proto supersedes all of this. Thoughts?

Reply:
I think JSON is useful either way. It's not very hard to read. (mtool just takes raw JSON and pretty prints it). 

I don't really care that much about whether we keep the table or not, but I don't think it's necessary even for human operators. It is not difficult to look at well formatted JSON.

If you want to keep the table and it's tested that's fine. I would still output the JSON, as long as it is valid JSON and can be formatted via mtool (which expects well formed JSON). We can swap with a proto later.

Reply:
Sorry in looking at this again briefly, is the JSON output not implemented yet? If not I would definitely implement it and just do something really easy/simple. Pick whatever format you want (personally, I would just do a a dictionary, where the key maps to an object that has the final value, as well as a list of composite layers). If you feel it's easier to do this via a proto vs. hand writing feel free.

I would still just not bother with the textual table output at all. I don't think it's worth supporting, but again, don't feel that strongly about it. We can see if @jmarantz or @htuch have any thoughts on that aspect.

Reply:
OK, I'll just implement minimally viable manual JSON and nuke the pretty table; that code was nasty to write and is nastier to read anyways.

Even though it will drag things out due to PR in dpapi, I will do proto for this. To set an example or whatever.